File: server/velvet_dawn/models/datapacks/tiles/tile_textures.py
import random
from typing import List, Optional

from velvet_dawn import errors


""" Tile textures data

Tiles must have a color but also a possible assignable overlay image.
A tile may be defined with a list of colors/images but only one will
be chosen when the tile is created.
"""


AVAILABLE_KEYS = {"color", "image", "notes"}


class TileTextures:
    def __init__(self):
        self.colors: List[str] = []
        self.images: List[str] = []

    def choose_color(self) -> Optional[str]:
        """ Choose the tile color

        Returns:
            The chosen color
        """
        return random.choice(self.colors)

    def choose_image(self) -> Optional[str]:
        """ Choose the image

        Returns:
            The chosen image
        """
        if not self.images:
            return None

        choice = random.choice(self.images)
        if choice == "null":
            return None

        return choice

    @staticmethod
    def load(tile_id: str, data: dict):
        """ Parse the data for load the colors and textures """
        textures = TileTextures()

        background_colors = data.get("color", "#0099ff")
        if isinstance(background_colors, str): textures.colors = [background_colors]
        elif isinstance(background_colors, list): textures.colors = background_colors
        elif isinstance(background_colors, dict):
            for key in background_colors:
                textures.colors += [key] * background_colors[key]
        else:
            raise errors.ValidationError(f"{tile_id} has invalid background textures.")

        background_textures = data.get("image")
        if isinstance(background_textures, str): textures.images = [background_textures]
        elif isinstance(background_textures, list): textures.images = background_textures
        elif isinstance(background_textures, dict):
            for key in background_textures:
                textures.images += [key] * background_textures[key]
        elif background_textures is None: pass
        else:
            raise errors.ValidationError(f"{tile_id} has invalid background textures.")

        # Check for random other keys
        for key in data:
            if key not in AVAILABLE_KEYS:
                raise errors.ValidationError(f"{tile_id} textures unknown key: '{key}'")

        return textures

File: src/jserver/src/velvetdawn/models/datapacks/tiles/TileTextures.java
package velvetdawn.models.datapacks.tiles;

import velvetdawn.models.anytype.Any;
import velvetdawn.models.anytype.AnyJson;
import velvetdawn.models.anytype.AnyList;
import velvetdawn.models.anytype.AnyString;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Set;

public class TileTextures {

    /* Tile textures data
    Tiles must have a color but also a possible assignable overlay image.
    A tile may be defined with a list of colors/images but only one will
    be chosen when the tile is created.
    */

    private static final Set<String> AvailableKeys = Set.of("color", "image", "notes");

    public List<String> colors = new ArrayList<>();
    public List<String> images = new ArrayList<>();

    /** Choose the tile colour */
    public Any chooseColor() {
        if (this.colors.isEmpty())
            return new AnyString( "#000000");

        Random rand = new Random();
        return new AnyString(this.colors.get(rand.nextInt(this.colors.size())));
    }

    /** Choose the image */
    public Any chooseImage() {
        if (images.isEmpty())
            return Any.Null();

        Random rand = new Random();
        String image = this.images.get(rand.nextInt(this.colors.size()));
        if (image.equals("null"))
            return Any.Null();

        return new AnyString(image);
    }

    /** Parse the data for load the colors and textures */
    public TileTextures load(String datapackId, AnyJson data) throws Exception {
        this.colors = parseTexturesType(datapackId, "color", data);
        this.images = parseTexturesType(datapackId, "image", data);

        // Check for random other keys
        for (String key: data.keys()) {
            if (!AvailableKeys.contains(key))
                throw new Exception(String.format("%s textures unknown key: '%s'", datapackId, key));
        }

        return this;
    }

    private List<String> parseTexturesType(String parentId, String key, AnyJson data) throws Exception {
        if (!data.containsKey(key))
            return List.of();

        Any value = data.get(key);

        if (value instanceof AnyString)
            return List.of(value.toString());

        if (value instanceof AnyList) {
            ArrayList<String> items = new ArrayList<>();
            for (Any item: ((AnyList) value).items) {
                items.add(item
                        .validateInstanceIsString(String.format("Error in %s. List of textures must be a string", parentId))
                        .value);
            }
            return items;
        }


        if (value instanceof AnyJson) {
            List<String> listItems = new ArrayList<>();

            for (String subkey : ((AnyJson) value).keys()) {
                float item = ((AnyJson) value).get(subkey)
                        .validateInstanceIsFloat(String.format("Count of %s %s must be a number in %s", key, subkey, parentId)).value;

                for (int i = 0; i < item; i++)
                    listItems.add(subkey);
            }

            return listItems;
        }

        throw new Exception(String.format("Malformed %s tag in %s", key, parentId));
    }
}
